File: src/main/python/link/impl/ReverseSingleLinkImpl.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @File    : ReverseSingleLinkImpl.py

class Node:
    data = 0
    next = None


def build_single_link_list(nodeArr):
    head_node = None
    next_node = None
    for i in nodeArr:
        node = Node()
        node.data = i
        node.next = None
        if head_node is None:
            head_node = node
            next_node = head_node
        else:
            next_node.next = node
            next_node = node

    return head_node


def reverse1(head):
    if head is None or head.next is None:
        return head
    current = head
    pre = None
    pnext = None
    while current is not None:
        pnext = current.next
        current.next = pre
        pre = current
        current = pnext

    return pre


def reverse2(current):
    if current.next is None:
        return current
    pnext = current.next
    current.next = None
    reversed = reverse2(pnext)
    pnext.next = current

    return reversed


def reverse3(current, pre):
    if current.next is None:
        current.next = pre
        return current
    else:
        pnext = current.next
        current.next = pre
        return reverse3(pnext, current)


if __name__ == '__main__':
    head = build_single_link_list([1, 2, 3, 4, 5])
    head = reverse1(head)
    print("impl1:")
    while head is not None:
        print(head.data)
        head = head.next
    print("impl2:")
    head = build_single_link_list([1, 2, 3, 4, 5])
    head = reverse2(head)
    while head is not None:
        print(head.data)
        head = head.next
    print("impl3:")
    head = build_single_link_list([1, 2, 3, 4, 5])
    head = reverse3(head,None)
    while head is not None:
        print(head.data)
        head = head.next

File: src/main/java/link/impl/ReverseSingleLinkImpl.java
package link.impl;

public class ReverseSingleLinkImpl {
    static class Node {
        int data;
        Node next;
    }

    public static void main(String[] args) {
        //reverse1
        System.out.println("impl1:");
        Node head = reverse1(buildSingleLinkList(new int[]{1, 2, 3, 4, 5}));
        while (head != null) {
            System.out.println(head.data);
            head = head.next;
        }
        //reverse2
        System.out.println("impl2:");
        head = reverse2(buildSingleLinkList(new int[]{1, 2, 3, 4, 5}));
        while (head != null) {
            System.out.println(head.data);
            head = head.next;
        }
        //reverse3
        System.out.println("impl3:");
        Node headNode = buildSingleLinkList(new int[]{1, 2, 3, 4, 5});
        head = reverse3(headNode, null);
        while (head != null) {
            System.out.println(head.data);
            head = head.next;
        }
    }

    /**
     * 非递归实现
     *
     * @param head
     * @return
     */
    static Node reverse1(Node head) {
        if (head == null || head.next == null) {
            return head;
        }
        Node current = head;
        Node pre = null;
        Node pnext = null;
        while (current != null) {
            pnext = current.next;
            current.next = pre;
            pre = current;
            current = pnext;
        }

        return pre;
    }

    /**
     * 递归实现
     *
     * @param current
     */
    static Node reverse2(Node current) {
        if (current.next == null) {
            return current;
        }
        Node pnext = current.next;
        current.next = null;
        Node reversed = reverse2(pnext);
        pnext.next = current;

        return reversed;
    }

    /**
     * @param current
     * @param pre
     * @return
     */
    static Node reverse3(Node current, Node pre) {
        if (current.next == null) {
            current.next = pre;
            return current;
        } else {
            Node next = current.next;
            current.next = pre;
            return reverse3(next, current);
        }
    }

    static Node buildSingleLinkList(int[] nodeArr) {
        Node head = null;
        Node next = null;
        for (int i : nodeArr) {
            Node node = new Node();
            node.data = i;
            node.next = null;
            if (head == null) {
                head = node;
                next = head;
            } else {
                next.next = node;
                next = node;
            }
        }
        return head;
    }
}
